zerver/tests/test_reactions.py

```diff
@@ -367,3 +368,462 @@ def test_remove_event(self):
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'green_tick',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_success(result)
+
+    def test_add_realm_emoji_invalid_code(self):
+        # type: () -> None
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'non_existent',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_error(result, 'Emoji for this emoji code not found.')
+
+    def test_add_deactivated_realm_emoji(self):
+        # type: () -> None
+        emoji = RealmEmoji.objects.get(name="green_tick")
+        emoji.deactivated = True
+        emoji.save(update_fields=['deactivated'])
+
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'green_tick',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_error(result, 'Emoji for this emoji code not found.')
+
+    def test_add_to_existing_deactivated_realm_emoji_reaction(self):
+        # type: () -> None
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'green_tick',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_success(result)
+
+        emoji = RealmEmoji.objects.get(name="green_tick")
+        emoji.deactivated = True
+        emoji.save(update_fields=['deactivated'])
+
+        result = self.post_reaction(reaction_info, sender='AARON')
+        self.assert_json_success(result)
+
+        reactions = self.get_message_reactions(1, 'green_tick', 'realm_emoji')
+        self.assertEqual(len(reactions), 2)
+
+    def test_remove_realm_emoji_reaction(self):
+        # type: () -> None
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'green_tick',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_success(result)
+
+        result = self.delete_reaction(reaction_info)
+        self.assert_json_success(result)
+
+    def test_remove_deactivated_realm_emoji_reaction(self):
+        # type: () -> None
+        reaction_info = {
+            'emoji_name': 'green_tick',
+            'emoji_code': 'green_tick',
+        }
+        result = self.post_reaction(reaction_info)
+        self.assert_json_success(result)
+
+        emoji = RealmEmoji.objects.get(name="green_tick")
+        emoji.deactivated = True
+        emoji.save(update_fields=['deactivated'])
+
+        result = self.delete_reaction(reaction_info)
+        self.assert_json_success(result)
+
+    def test_remove_non_existent_realm_emoji_reaction(self):
+        # type: () -> None
+        reaction_info = {
+            'emoji_name': 'non_existent',
+            'emoji_code': 'TBD',
+        }
+        result = self.delete_reaction(reaction_info)
+        self.assert_json_error(result, "Reaction doesn't exist.")
```

Comment:
I added one more test here to cover the case of an invalid `reaction_type`, which was exposed as needing coverage when I refactored `zerver/lib/emoji.py`.